### PR TITLE
Closes #2412 and #2314 - improve query on plugin home page

### LINKED ISSUE #2412: Plugin main page is very long to display
Currently, on my test server the plugin main page need **almost 30 seconds** to get displayed.

I found out that this is because of a very huge and not optimized SQL query built by the `Task::getJobLogs` function.

The SQL request:
```
SELECT task.`id` AS 'task.id', task.`name` AS 'task.name', job.`id` AS 'job.id', job.`name` AS 'job.name', job.`method` AS 'job.method', agent.`id` AS 'agent.id', agent.`name` AS 'agent.name', agent.`computers_id` AS 'agent.computers_id', run.`id` AS 'run.id', run.`itemtype` AS 'run.itemtype', run.`items_id` AS 'run.items_id', run.`state` AS 'run.state', log.`date` AS 'log.last_date', UNIX_TIMESTAMP(log.`date`) AS 'log.last_timestamp', log.`id` AS 'log.last_id', log.`comment` AS 'log.last_comment'
FROM `glpi_plugin_fusioninventory_agents` AS agent
INNER JOIN ( SELECT MAX(run.`id`) AS max_id, run.`plugin_fusioninventory_agents_id`, run.`plugin_fusioninventory_taskjobs_id`, run.`items_id`, run.`itemtype`, MAX(log.`id`) AS max_log_id
FROM ( SELECT *
FROM ( SELECT *, @num := IF(@agent_id = plugin_fusioninventory_agents_id && @taskjob_id = plugin_fusioninventory_taskjobs_id && @items_id = items_id && @itemtype = itemtype, @num:= @num + 1, 1) AS row_num, @agent_id:=plugin_fusioninventory_agents_id AS tmp_var1, @taskjob_id:=plugin_fusioninventory_taskjobs_id AS tmp_var2, @items_id:=items_id AS tmp_var3, @itemtype:=itemtype AS tmp_var4
FROM glpi_plugin_fusioninventory_taskjobstates
ORDER BY plugin_fusioninventory_taskjobs_id, plugin_fusioninventory_agents_id, itemtype, items_id, id DESC ) AS limited_states
WHERE row_num <= 2 ) AS run
LEFT JOIN `glpi_plugin_fusioninventory_taskjoblogs` AS log ON log.`plugin_fusioninventory_taskjobstates_id` = run.`id` GROUP BY run.`plugin_fusioninventory_agents_id`, run.`plugin_fusioninventory_taskjobs_id`, run.`items_id`, run.`itemtype`, run.`id` ) max_run ON max_run.`plugin_fusioninventory_agents_id` = agent.`id`
INNER JOIN `glpi_plugin_fusioninventory_taskjobstates` AS run ON max_run.`max_id` = run.`id`
LEFT JOIN `glpi_plugin_fusioninventory_taskjoblogs` as log ON max_run.`max_log_id` = log.`id`
INNER JOIN `glpi_plugin_fusioninventory_taskjobs` AS job ON job.`id` = run.`plugin_fusioninventory_taskjobs_id`
INNER JOIN `glpi_plugin_fusioninventory_tasks` as task ON job.`plugin_fusioninventory_tasks_id` = task.`id`
WHERE 1 AND task.`id` IN (0,6,12,13,14,15) GROUP BY job.`id`, agent.`id`, run.`id`, log.`id`
ORDER BY agent.`id` ASC, run.`id` DESC;
```

The EXPLAIN of this request:
image

I am not enough SQL aware to propose an optimization of this request 😞 Anyone ?

### PR DESCRIPTION
**Pull request by mohierf**
Improve the huge query used to get the tasks jobs log (closes #2412)
Add some extra-debug logs for this process
Allow to get information with or without execution result log
Allow to view graph/log for non active tasks (closes #2314)

### CODE REVIEW

File: inc/task.class.php [hunk 1]
```diff
@@ -944,108 +969,113 @@ function getJoblogs($task_ids = []) {
...
-         $query_select[] = $key[1] . " AS '" . $key[0] . "'";
-      }
-
-      $query_joins         = [];
-      $query_joins['task'] = "
-         INNER JOIN `glpi_plugin_fusioninventory_tasks` as task
-            ON job.`plugin_fusioninventory_tasks_id` = task.`id`
-            AND task.`is_active` = 1";
-
+      $active_task = $only_active ? "AND `is_active`='1'" : "";
```
Comment (wawax) [inline inc/task.class.php]: Please replace SQL queries by DBIterator

File: inc/task.class.php [hunk 2]
```diff
@@ -944,108 +969,113 @@ function getJoblogs($task_ids = []) {
...
-         $task_id = $result[$fieldmap['task.id']];
+      $agent_state_types = [
+         'agents_prepared',
+         'agents_cancelled',
+         'agents_running',
+         'agents_success',
+         'agents_error',
+         'agents_notdone'
+      ];
+      while ($result = $data_structure['result']->fetch_assoc()) {
```
Comment (wawax) [inline inc/task.class.php]: Use foreach instead of while

File: inc/task.class.php [hunk 3]
```diff
@@ -1073,308 +1103,293 @@ function getJoblogs($task_ids = []) {
...
+      $q_task_job_state['result'] = $DB->query($q_task_job_state['query']);
+      $q_task_job_state['end'] = microtime(true);
+      $q_task_job_state['duration'] = self::formatChrono($q_task_job_state);
+      PluginFusioninventoryToolbox::logIfExtradebug(
+         "pluginFusioninventory-tasks",
+         $q_task_job_state);
+
+      // Parse the query result to update the data to return
+      $count_results = 0;
+      while ($result = $q_task_job_state['result']->fetch_assoc()) {
```
Comment (wawax) [inline inc/task.class.php]: please use foreach instead of while + fetch_assoc

File: inc/task.class.php [hunk 4]
```diff
@@ -1073,308 +1103,293 @@ function getJoblogs($task_ids = []) {
...
+                  'last_log_id' => $log_result['log.last_id'],
+                  'last_log_date' => $log_result['log.last_date'],
+                  'timestamp' => $log_result['log.last_timestamp'],
+                  'last_log' => $log_result['log.last_comment']
                ];
             }
+         } else {
+            $targets[$target_id]['agents'][$agent_id][] = [
+               'agent_id' => $agent_id,
+               'link' => $CFG_GLPI['root_doc'] . "/front/computer.form.php?id=" . $result['agent.computers_id'],
```
Comment (wawax) [inline inc/task.class.php]: Please use Computer::getFormURLWithID()

**Comment by mohierf** [comment]
@wawax this PR is a backport of #2435 on the master branch ... it would be better to update #2435 and then, once merged, backport the modifications on the master branch 😉

**Pull request closed without merging**